### PR TITLE
Ability to disable Identify for a layer #6101

### PR DESCRIPTION
**Pull request by myrdstom**
## Description

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#6101

**What is the new behavior?**

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information

### CODE REVIEW

**Review by offtherailz — CHANGES_REQUESTED** [review]
image
I don't think this is the desired behavior. 
I think that the request should not be performed at all, and nothing should be shown, filtering out the layer from the feature info list, as well as the one that not support it (e.g. TMS ...).

Correct me if I'm wrong @tdipisa 

Moreover, here I have a doubt about the "Feature info only on selected layer" what the expected behavior if the selected layer is disabled ? 
I think the usual no info layer modal should be shown. Can you confirm @tdipisa

**Comment by myrdstom** [comment]
Yes, it isn't, I will be adjusting as specified

**Comment by tdipisa** [comment]
@offtherailz yes, I've already discussed your comment with @myrdstom. He is working to complete the implementation.

And yes, in the following case:

>Moreover, here I have a doubt about the "Feature info only on selected layer" what the expected behavior if the selected layer is disabled ?

It must be as you have reported:

>I think the usual no info layer modal should be shown. Can you confirm @tdipisa

File: web/client/epics/identify.js [hunk 1]
```diff
@@ -162,6 +162,13 @@ export const getFeatureInfoOnFeatureInfoClick = (action$, { getState = () => { }
                         const itemId = itemIdSelector(getState());
                         const reqId = uuid.v1();
                         const param = { ...appParams, ...requestParams };
+                        const infoFormat = getState()?.layers?.flat[6]?.featureInfo?.format;
```
Comment (offtherailz) [inline web/client/epics/identify.js:165]: This is uncorrect, here you are getting for every layer the state of the 6th layer. Moreover, you should not create a request for this layer, anyway. You can filter them on queraible layer selector (line 125, 126), and there is already a flag `queryable` for this purpose. So I think you don't have to touch this file, but simply, from the UI, set the queryable flag of the layer, when the format is selected.

File: web/client/utils/FeatureInfoUtils.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 
 const INFO_FORMATS = {
+    "HIDDEN": "text/html",
```
Comment (offtherailz) [inline web/client/utils/FeatureInfoUtils.js:3]: I don't think this is needed

File: web/client/utils/MapInfoUtils.js [hunk 3]
```diff
@@ -17,7 +17,7 @@ const MapInfoUtils = {
      * specifies which info formats are currently supported
      */
     //           default format ↴
-    AVAILABLE_FORMAT: ['TEXT', 'PROPERTIES', 'HTML', 'TEMPLATE'],
+    AVAILABLE_FORMAT: ['HIDDEN', 'TEXT', 'PROPERTIES', 'HTML', 'TEMPLATE'],
```
Comment (offtherailz) [inline web/client/utils/MapInfoUtils.js:20]: I don't think this is not needed

**Review by offtherailz — CHANGES_REQUESTED** [review]
The current solution is not correct (you hardcoded to get the only the 7th layer). I think you should move your check in a different place.
Please see comments inline.
Moreover:
- [ ] unit tests should be provided
- [ ] The icon should be different
image
- [ ] I think the info format is not the proper value to store in the layer. Internally (see MapInfoUtils line 213) there is already a flag, `queryable` that provides this functionality. You should instead set that flag when the user select "HIDDEN" to disable info.

File: web/client/epics/identify.js (hunk 1 above)
Comment (myrdstom) [reply web/client/epics/identify.js:165]: I must have pushed this by mistake. I have readjusted to: queryableLayers && queryableLayers[0] && queryableLayers[0].featureInfo && queryableLayers[0].featureInfo.format;

File: web/client/utils/MapInfoUtils.js (hunk 3 above)
Comment (myrdstom) [reply web/client/utils/MapInfoUtils.js:20]: Screenshot 2020-11-04 at 13 14 31

DefaultSettings calls the function `MapInfoUtils.getAvailableInfoFormat()` which determines what settings are available in TOC, is there another way to go about it?

File: web/client/epics/identify.js (hunk 1 above)
Comment (myrdstom) [reply web/client/epics/identify.js]: With regards the `DISABLE IDENTITY` icon being different, I had a brief chat with @tdipisa and he mentioned he would discuss it... I can't recall whom exactly but I think you or Stefano. I can raise it with him again

File: web/client/utils/MapInfoUtils.js (hunk 3 above)
Comment (offtherailz) [reply web/client/utils/MapInfoUtils.js:20]: `HIDDEN` is not a real featureInfo  format, so I think you should handle the accordion to set/unset properly the `queryable` flag, and keeping the feature info format unchanged, as well as the utilities dedicated to it.
You can append it here, as a fake info format, for the accordion. Then when click, check it and set/unset queryable accordingly, or set the current info format.

File: web/client/epics/identify.js (hunk 1 above)
Comment (tdipisa) [reply web/client/epics/identify.js]: @offtherailz I will involve Stefano to provide a better icon for this as soon as the PR is merged.

File: web/client/utils/MapInfoUtils.js (hunk 3 above)
Comment (myrdstom) [reply web/client/utils/MapInfoUtils.js]: So what I have done is append `HIDDEN` to the result of `getAvailableInfoFormat()`. I also Append it to the MapInfoUtils.INFO_FORMATS as that is used throughout the file

**Comment by MV88** [comment]
@myrdstom  a small suggestion: avoid to use so much force push on a created pr, it will be hard to identify the changes between commits.

do that locally before creating a pr, but once ti is created is better to avoid that

**Comment by myrdstom** [comment]
> @myrdstom a small suggestion: please, avoid to use so much force push on a created pr, it will be hard to identify the changes between commits.
> 
> maybe you can do that locally before creating a pr, but once it is created is better to avoid that operation
> commits description should describe what the commit does in a few words

Sorry for the new commit mess... Early morning. This won't happen moving forward

**Review by offtherailz — CHANGES_REQUESTED** [review]
I think you committed the first solution. (both behavior and files are very similar)
Are you still working on it?

**Comment by myrdstom** [comment]
> I think you committed the first solution. (both behavior and files are very similar)
> Are you still working on it?

No, It's ready for review

File: web/client/epics/identify.js [hunk 4]
```diff
@@ -162,6 +162,13 @@ export const getFeatureInfoOnFeatureInfoClick = (action$, { getState = () => { }
                         const itemId = itemIdSelector(getState());
                         const reqId = uuid.v1();
                         const param = { ...appParams, ...requestParams };
+                        const infoFormat = queryableLayers[0]?.featureInfo?.format;
+                        if (infoFormat === 'HIDDEN') {
+                            return [
+                                newMapInfoRequest(reqId, param),
+                                loadFeatureInfo(reqId, [{}], requestParams, { ...lMetaData, features: 0, featuresCrs: `EPSG:4326` }, layer)
+                            ];
+                        }
```
Comment (offtherailz) [inline web/client/epics/identify.js:171]: filter the layers at lines 125-126

File: web/client/components/data/identify/viewers/HiddenViewer.jsx [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+            <div id="container">
+                <Alert bsStyle={"danger"}>
+                    <h4><HTML msgId="noFeatureInfo"/></h4>
+                </Alert>
+            </div>
+        );
+    }
+}
+
+module.exports = HiddenViewer;
```
Comment (offtherailz) [inline web/client/components/data/identify/viewers/HiddenViewer.jsx:19]: hidden viewer and it's test are not needed anymore

**Pull request closed without merging**